moshi/src/main/java/com/squareup/moshi/internal/Util.java

```diff
@@ -520,4 +531,91 @@ public static String typeAnnotatedWithAnnotations(Type type,
+      throw new IllegalStateException("DefaultConstructorMarker not on classpath. Make sure the "
+          + "Kotlin stdlib is on the classpath.");
+    }
+    Constructor<T> defaultConstructor = findConstructor(targetClass);
+    defaultConstructor.setAccessible(true);
+    return defaultConstructor;
+  }
+
+  /**
+   * Reflectively invokes the defaults constructor of a kotlin class. This allows indicating which
+   * arguments are "set" or not, and thus recreate the behavior of named a arguments invocation
+   * dynamically.
+   *
+   * @param targetClass the target kotlin class to instantiate.
+   * @param defaultsConstructor the target class's defaults constructor in kotlin invoke.
+   * @param mask an int mask indicating which {@code args} are present.
+   * @param args the constructor arguments, including "unset" values (set to null or the primitive
+   *             default).
+   * @param <T> the type of {@code targetClass}.
+   * @return the instantiated {@code targetClass} instance.
+   * @see #createDefaultValuesParametersMask(boolean...)
+   */
+  public static <T> T invokeDefaultConstructor(
+      Class<T> targetClass,
+      Constructor<T> defaultsConstructor,
+      int mask,
+      Object... args) {
+    Object[] finalArgs = Arrays.copyOf(args, args.length + 2);
+    finalArgs[finalArgs.length - 2] = mask;
+    finalArgs[finalArgs.length - 1] = null; // DefaultConstructorMarker param
+    try {
+      return defaultsConstructor.newInstance(finalArgs);
+    } catch (InstantiationException e) {
+      throw new IllegalStateException("Could not instantiate instance of " + targetClass);
+    } catch (IllegalAccessException e) {
+      throw new IllegalStateException("Could not access defaults constructor of " + targetClass);
+    } catch (InvocationTargetException e) {
+      Throwable cause = e.getCause();
+      if (cause instanceof RuntimeException) throw (RuntimeException) cause;
+      if (cause instanceof Error) throw (Error) cause;
+      throw new RuntimeException("Could not invoke defaults constructor of " + targetClass, cause);
+    }
+  }
+
+  private static <T> Constructor<T> findConstructor(Class<T> targetClass) {
+    for (Constructor<?> constructor : targetClass.getDeclaredConstructors()) {
+      Class<?>[] paramTypes = constructor.getParameterTypes();
+      if (paramTypes.length != 0
+          && paramTypes[paramTypes.length - 1].equals(DEFAULT_CONSTRUCTOR_MARKER)) {
+        //noinspection unchecked
+        return (Constructor<T>) constructor;
+      }
+    }
+
+    throw new IllegalStateException("No defaults constructor found for " + targetClass);
+  }
+
+  /**
+   * Creates an mask with bits set to indicate which indices of a default constructor's parameters
+   * are set.
+   *
+   * @param argPresentValues vararg of all present values (set or unset). Max allowable size is 32.
+   * @return the created mask.
+   */
+  public static int createDefaultValuesParametersMask(boolean... argPresentValues) {
```

Comment:
I vaguely remember commenting about moving this logic to the generated code which would eliminate the "is present" booleans and avoid the boolean array allocation to call this method. Was there a reason that didn't happen?

Reply:
I misunderstood and thought you were just asking to make this argument a vararg. You think we should inline _all_ of this? It'd be a potentially really long cascade of if-checks

Reply:
Could the mask itself move out to the adapter, then instead of property and propertySet, you'd just update the mask int when the value is read, right inside the when?

Reply:
Right I'm saying we don't use booleans and just use the int and set bits inside the generated code.

Reply:
I think I understand what mean now 👍. Will do but in a separate PR from the others above since it's a little more involved

Reply:
Note we can't entirely escape local booleans, but now they'd only be used for gating setting non-constructor `var` properties

Reply:
This gets a little tricky, because the mask only sets bits for parameters that are _not_ set. This means we'd likely have to manually set every bit first, and then unset them as we parse types. Worth the overhead?

Reply:
Is it a problem to just invert the mask?

```kotlin
var mask = 1
when (reader.selectName(options)) {
                0 -> {
                    nullableFirstName = stringAdapter.fromJson(reader)
                    mask = mask and (0 shl 0)
                }
                1 -> {
                    nullableLastName = stringAdapter.fromJson(reader)
                    mask = mask and (0 shl 1)
                } 
                ...
```

Reply:
Nope that works, I had tried that and got test failures and assumed it was wrong. Turned out it was right, and something else had broken. Teaches me to read the test output better :)

Opened a PR: #908